### PR TITLE
BUG: Should not raises errors in .set_names and comparison methods for MultiIndex with nlevels == 1 (GH21149)

### LINKED ISSUE #21149: (Assuming we don't want to disallow 1-level MultiIndexes), stop identifying MultiIndex by nlevels == 1
#### Code Sample, a copy-pastable example if possible

```python
In [2]: mi = pd.MultiIndex.from_product([[0, 1]])

In [3]: mi.set_names(('first',), level=0)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-1f4ad6c7f0bf> in <module>()
----> 1 mi.set_names(('first',), level=0)

/home/nobackup/repo/pandas/pandas/core/indexes/base.py in set_names(self, names, level, inplace)
   1386 
   1387         if level is not None and self.nlevels == 1:
-> 1388             raise ValueError('Level must be None for non-MultiIndex')
   1389 
   1390         if level is not None and not is_list_like(level) and is_list_like(

ValueError: Level must be None for non-MultiIndex

```

#### Problem description

The case of ``MultiIndex`` with ``self.nlevels == 1`` is currently perfectly valid (despite e.g. ``MultiIndex.droplevel`` returning a normal ``Index`` when only one level is left). Unless we want to change this, we should always identify a ``MultiIndex`` with ``isinstance``, not by looking at ``self.nlevels``.

#### Expected Output

```python
In [4]: mi.set_names(('first',))
Out[4]: 
MultiIndex(levels=[[0, 1]],
           labels=[[0, 1]],
           names=['first'])
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: 5cf4957a2cc8429170742dd1e6435ddb8c323811
python: 3.5.3.final.0
python-bits: 64
OS: Linux
OS-release: 4.9.0-6-amd64
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: it_IT.UTF-8
LOCALE: it_IT.UTF-8

pandas: 0.24.0.dev0+14.g5cf4957a2.dirty
pytest: 3.5.0
pip: 9.0.1
setuptools: 39.0.1
Cython: 0.25.2
numpy: 1.14.3
scipy: 0.19.0
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: 1.5.6
patsy: 0.5.0
dateutil: 2.7.3
pytz: 2018.4
blosc: None
bottleneck: 1.2.0dev
tables: 3.3.0
numexpr: 2.6.1
feather: 0.3.1
matplotlib: 2.0.0
openpyxl: 2.3.0
xlrd: 1.0.0
xlwt: 1.3.0
xlsxwriter: 0.9.6
lxml: 4.1.1
bs4: 4.5.3
html5lib: 0.999999999
sqlalchemy: 1.0.15
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: 0.2.1

</details>

### PR DESCRIPTION
**Pull request by KalyanGokhale**
- [x] closes #21149 
- [ ] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by KalyanGokhale** [comment]
@toobaz have made necessary updates for #21149 

original issue with set_names is fixed

Also, with the previous code snippet in `def cmp_method` it raised `ValueError` for comparisons of MultiIndex with nlevels == 1, which was inconsistent with behaviour for MultiIndex with nlevels > 1 - this has now been fixed

there is some conflict with file `whatsnew/v0.23.1.txt` that I was not able to resolve - and it is failing the `continuous-integration/appveyor/pr` check - please can you help resolve it?

**Comment by KalyanGokhale** [comment]
Resolved the earlier conflict with `whatsnew` file

**Comment by KalyanGokhale** [comment]
Will check the test failure and resubmit after necessary edits

File: pandas/tests/indexes/test_multi.py [hunk 1]
```diff
@@ -3291,3 +3291,37 @@ def test_duplicate_multiindex_labels(self):
         with pytest.raises(ValueError):
             ind.set_levels([['A', 'B', 'A', 'A', 'B'], [2, 1, 3, -2, 5]],
                            inplace=True)
+
+    testdata = [
+        (pd.MultiIndex.from_product([[0, 1], [1, 0]]), pd.Series(range(4)), 4),
```
Comment (jreback) [inline pandas/tests/indexes/test_multi.py]: don't create a variable here, just put it directly in the parametrize (see how to format it from other tests)

File: pandas/tests/indexes/test_multi.py [hunk 2]
```diff
@@ -3291,3 +3291,37 @@ def test_duplicate_multiindex_labels(self):
...
+        (pd.MultiIndex.from_product([[0, 1]]), pd.Series(range(2)), 2),
+    ]
+
+    @pytest.mark.parametrize("midx,idx,count", testdata)
+    def test_multiindex_compare(self, midx, idx, count):
+        # GH 21149 - change in 'def cmp_method()'
```
Comment (jreback) [inline pandas/tests/indexes/test_multi.py]: this is a non-sensical comment, be more descriptive.

File: pandas/tests/indexes/test_multi.py [hunk 3]
```diff
@@ -3291,3 +3291,37 @@ def test_duplicate_multiindex_labels(self):
...
+        expected.reset_index(drop=True, inplace=True)
+        result = pd.Series(midx > midx)
+        tm.assert_series_equal(result, expected)
+        result = pd.Series(midx == idx)
+        tm.assert_series_equal(result, expected)
+
+        with tm.assert_raises_regex(TypeError, 'not supported'):
+            midx > idx
+
+    def test_multiindex_set_names(self):
```
Comment (jreback) [inline pandas/tests/indexes/test_multi.py]: put this near the other tests for set_names

File: pandas/tests/indexes/test_multi.py [hunk 4]
```diff
@@ -3291,3 +3291,37 @@ def test_duplicate_multiindex_labels(self):
...
+        (pd.MultiIndex.from_product([[0, 1], [1, 0]]), pd.Series(range(4)), 4),
+        (pd.MultiIndex.from_product([[0, 1]]), pd.Series(range(2)), 2),
+    ]
+
+    @pytest.mark.parametrize("midx,idx,count", testdata)
+    def test_multiindex_compare(self, midx, idx, count):
+        # GH 21149 - change in 'def cmp_method()'
+        expected = pd.Series([True]).repeat(count)
+        expected.reset_index(drop=True, inplace=True)
+        result = pd.Series(midx == midx)
```
Comment (jreback) [inline pandas/tests/indexes/test_multi.py]: not really clear what you are testing, this is way too dense. put blank lines and a comment in between cases.

**Review by jreback — CHANGES_REQUESTED** [review]
you are doing 2 different changes here which are unreleated, break this in to 2 PRs

**Comment by KalyanGokhale** [comment]
@jreback Thanks for the review - have split this into two PRs (#21195 and #21196). Have also made the suggested edits. Closing this PR.

**Pull request closed without merging**